File: src/visualizer.py
from __future__ import annotations

import os
import subprocess
import shutil

from loguru import logger


def _find_script(name: str) -> str:
    return shutil.which(name) or os.path.join(os.path.dirname(__file__), "..", "scripts", name)


def check_tools_available() -> dict[str, bool]:
    png_tool = shutil.which("gltf_to_png")
    webm_tool = shutil.which("gltf_to_webm")
    return {
        "gltf_to_png": bool(png_tool),
        "gltf_to_webm": bool(webm_tool),
    }


def generate_png_preview(scene_gltf_path: str, output_path: str) -> str:
    script = _find_script("gltf_to_png")
    logger.info("Rendering PNG: {} -> {}", os.path.basename(scene_gltf_path), os.path.basename(output_path))
    result = subprocess.run(
        [script, "-i", scene_gltf_path, "-o", output_path],
        capture_output=True, text=True, timeout=120,
    )
    if result.returncode != 0:
        raise RuntimeError(f"gltf_to_png failed: {result.stderr}")
    return output_path


def generate_webm_preview(scene_gltf_path: str, output_path: str) -> str:
    script = _find_script("gltf_to_webm")
    logger.info("Rendering WebM: {} -> {}", os.path.basename(scene_gltf_path), os.path.basename(output_path))
    result = subprocess.run(
        [script, "-i", scene_gltf_path, "-o", output_path],
        capture_output=True, text=True, timeout=300,
    )
    if result.returncode != 0:
        raise RuntimeError(f"gltf_to_webm failed: {result.stderr}")
    return output_path


def generate_intermediate_frames(scene_gltf_path: str, output_dir: str) -> list[str]:
    os.makedirs(output_dir, exist_ok=True)
    preview_path = os.path.join(output_dir, "intermediate_preview.webm")
    generate_webm_preview(scene_gltf_path, preview_path)
    return [preview_path]


def generate_final_previews(scene_path: str, output_dir: str) -> dict[str, str]:
    logger.info("Generating final previews in {}", output_dir)
    os.makedirs(output_dir, exist_ok=True)
    png_path = os.path.join(output_dir, "preview.png")
    webm_path = os.path.join(output_dir, "preview.webm")
    results = {}
    if check_tools_available().get("gltf_to_png"):
        try:
            results["png"] = generate_png_preview(scene_path, png_path)
        except RuntimeError:
            results["png"] = ""
    if check_tools_available().get("gltf_to_webm"):
        try:
            results["webm"] = generate_webm_preview(scene_path, webm_path)
        except RuntimeError:
            results["webm"] = ""
    return results

File: src/visualizer.rs
use std::collections::HashMap;
use std::path::Path;
use std::process::Command;

use log::{info, warn};

pub fn check_tools_available() -> HashMap<String, bool> {
    let mut result = HashMap::new();
    result.insert(
        "gltf_to_png".to_string(),
        which("gltf_to_png").is_some(),
    );
    result.insert(
        "gltf_to_webm".to_string(),
        which("gltf_to_webm").is_some(),
    );
    result
}

fn which(name: &str) -> Option<String> {
    std::env::var("PATH").ok().and_then(|path| {
        for dir in path.split(':') {
            let full = Path::new(dir).join(name);
            if full.exists() {
                return Some(full.to_string_lossy().to_string());
            }
        }
        None
    })
}

pub fn generate_png_preview(scene_gltf_path: &str, output_path: &str) -> Result<String, String> {
    let script = which("gltf_to_png")
        .ok_or_else(|| "gltf_to_png not found on PATH".to_string())?;
    info!(
        "Rendering PNG: {} -> {}",
        Path::new(scene_gltf_path)
            .file_name()
            .map(|n| n.to_string_lossy())
            .unwrap_or_default(),
        Path::new(output_path)
            .file_name()
            .map(|n| n.to_string_lossy())
            .unwrap_or_default(),
    );
    let output = Command::new(&script)
        .args(["-i", scene_gltf_path, "-o", output_path])
        .output()
        .map_err(|e| format!("Failed to run gltf_to_png: {}", e))?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("gltf_to_png failed: {}", stderr));
    }
    Ok(output_path.to_string())
}

pub fn generate_webm_preview(scene_gltf_path: &str, output_path: &str) -> Result<String, String> {
    let script = which("gltf_to_webm")
        .ok_or_else(|| "gltf_to_webm not found on PATH".to_string())?;
    info!(
        "Rendering WebM: {} -> {}",
        Path::new(scene_gltf_path)
            .file_name()
            .map(|n| n.to_string_lossy())
            .unwrap_or_default(),
        Path::new(output_path)
            .file_name()
            .map(|n| n.to_string_lossy())
            .unwrap_or_default(),
    );
    let output = Command::new(&script)
        .args(["-i", scene_gltf_path, "-o", output_path])
        .output()
        .map_err(|e| format!("Failed to run gltf_to_webm: {}", e))?;
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("gltf_to_webm failed: {}", stderr));
    }
    Ok(output_path.to_string())
}

pub fn generate_intermediate_frames(
    scene_gltf_path: &str,
    output_dir: &str,
) -> Result<Vec<String>, String> {
    std::fs::create_dir_all(output_dir)
        .map_err(|e| format!("Failed to create output dir: {}", e))?;
    let preview_path = format!("{}/intermediate_preview.webm", output_dir);
    generate_webm_preview(scene_gltf_path, &preview_path)?;
    Ok(vec![preview_path])
}

pub fn generate_final_previews(
    scene_path: &str,
    output_dir: &str,
) -> HashMap<String, String> {
    info!("Generating final previews in {}", output_dir);
    let _ = std::fs::create_dir_all(output_dir);
    let png_path = format!("{}/preview.png", output_dir);
    let webm_path = format!("{}/preview.webm", output_dir);
    let mut results = HashMap::new();

    let tools = check_tools_available();
    if *tools.get("gltf_to_png").unwrap_or(&false) {
        match generate_png_preview(scene_path, &png_path) {
            Ok(p) => {
                results.insert("png".to_string(), p);
            }
            Err(e) => {
                warn!("PNG preview failed: {}", e);
                results.insert("png".to_string(), String::new());
            }
        }
    }
    if *tools.get("gltf_to_webm").unwrap_or(&false) {
        match generate_webm_preview(scene_path, &webm_path) {
            Ok(w) => {
                results.insert("webm".to_string(), w);
            }
            Err(e) => {
                warn!("WebM preview failed: {}", e);
                results.insert("webm".to_string(), String::new());
            }
        }
    }
    results
}
